### DISCUSSION TITLE
How are waveforms in CellExplorer being extracted?

### DISCUSSION THREAD

**Discussion by ajcheng1** [petersenpeter/CellExplorer#123] [Q&A] [answered]
I have a binary dat file that I sorted with kilosort and then manually curated with phy. just as a test, I labeled all the waveform units as good and it saved in cluster_group.tsv as good. But CellExplorer kept saying all the channels were recorded as bad. 
I realize CellExplorer was extracting the bad channels from getBadChannelsFromDat.m. When I comment several lines of code from that matlab file (line 97-100) to open the CellExplorer GUI, the waveforms sorted looked nothing alike to what was sorted and displayed in kilosort. Are the waveforms being extracted straight from the dat file and not the sorted spike data? 

%newBadChannels = find(wfF_std2 > params.noiseStdThreshold);
%newBadChannels2 = find(wfF_range > params.noiseRangeThreshold);
%newBadChannels = unique([newBadChannels,newBadChannels2]);
%session.channelTags.Bad.channels = unique([badChannels,newBadChannels]);

**Comment by petersenpeter** [comment, marked as answer]
Hi 

The waveforms are indeed extracted from the binary data. This is to make them consistent among users and various pipelines. By default, the pipeline will try to detect noisy/bad channels. This feature can be turned off, but it matters for the estimation of the peak spike waveform across your channels: https://github.com/petersenpeter/CellExplorer/blob/main/calc_CellMetrics/getWaveformsFromDat.m#L61

Here is the function detecting the bad channels: https://github.com/petersenpeter/CellExplorer/blob/main/calc_CellMetrics/getBadChannelsFromDat.m

Are you saying that all your channels are labeled as bad? If so, you could have a scaling issue.

**Reply by ajcheng1** [reply]
Yep, you are right! I was using int16 instead of single since my .dat file was in float32. However, the waveforms still don't look right, but it may be how I set up my .dat format. 

On a side note, I am using kilosort4 which doesn't use rez.mat anymore. Do you suggest I just manually input the electrode features?

**Reply by petersenpeter** [reply]
We are working on an implementation for Kilosort4. Let me get back to you on this. I believe we can get the electrode features from phy instead.

You can find more information on the format here: https://cellexplorer.org/datastructure/data-structure-and-format/#raw-data-file-format. The simplest test is to open your data with NeuroScope2: https://cellexplorer.org/interface/neuroscope2/

**Reply by ajcheng1** [reply]
Is there a way to extract waveforms from phy and not from dat? 
<img width="366" alt="image" src="https://github.com/user-attachments/assets/35bc58f6-2e6d-496c-81c2-2d3338ef444b" />
The waveforms present do not resemble what I see when I curate it with Phy.

I did specify the numeric type to be single, but does the numeric type impact the quality of the waveform appearance? 
```matlab
session.extracellular.precision = 'single';  % Numeric data type

```

**Reply by petersenpeter** [reply]
Alvin, I would be happy to try to troubleshoot it if you shared a session with me? If Kilosort can read your data, so can CellExplorer. 

If so I would need the binary data, relevant metadata and your spike sorted output from phy.

**Reply by ajcheng1** [reply]
I sent the files to your email address. Thanks!

**Reply by petersenpeter** [reply]
I think your spike times are not correct. Perhaps your sampling rate is not 12500Hz, or you have a different binary file for Kilosort. Your spike times do not align with any spikes in the traces. Your precision is set to 1, which is likely also incorrect. Finally, I cannot make sense of your probe layout, but I don't know what your probe looks like.

**Reply by ajcheng1** [reply]
Thanks for taking a look at it. I verified that the sampling rate on the MEA plate is 12500 hz, so it is likely due to spike timing. I think it might have to do with having to flattened the .dat file. 

The original dat file that was used for kilosort (Well A1.dat) was loaded from a matlab code. The script converted the data into dat so its data matrix was samples/time x channels. I flattened the .dat file and loaded it to CellExplorer only (was not used for kilosort) which probably affected the timing.

How were you able to check the spike times and the traces? I will do some debugging to look into this further. 

Here is what the probe looks like. Not too sure where I will be able to find the least significant bit unfortunately as the producer does not mention anything about it. 
![image](https://github.com/user-attachments/assets/6d3f70ae-c60d-41bb-991f-e662a7f562e6)

**Reply by ajcheng1** [reply]
Hey Peter, 
I checked and the spike timing was the issue. 
I got it to work and the spikes look good! 
<img width="1119" alt="image" src="https://github.com/user-attachments/assets/cc190c9e-1706-45af-9cd1-8cb09d265b5b" />
Much appreciation for the help with the debugging 

edit: I used 0.195 as the least significant bit. I am not to sure whether that will affect the neuron classification

**Reply by petersenpeter** [reply]
Great! See if you can increase your sampling rate to 20kHz or higher in the future. It should improve your spike sorting and waveform estimation. 

A sidenote: The spike waveform amplitude is too low by a factor of 10 or more. I would suggest you revisit the scaling (precision and least significant bit in CellExplorer).
